### ISSUE TITLE
Define and use a Run data model

### ISSUE THREAD

**Issue by decentralion** [tensorflow/tensorboard#581]
# Overview
"Runs" are one of the fundamental units of data organization in TensorBoard. They are the primary way of compartmentalizing data across different invocations of TensorFlow and diffing the results in the UI.

Surprisingly, given their privileged role in the system, they have no explicit data model. Instead, runs are defined only by their location on disk, and the only metadata available is the directory name. Consequently, managing TensorBoards with many different runs is quite cumbersome. 

As we move into the TensorBase/SQL database world, we can reasonably expect both that the number of runs available will dramatically increase (increasing the need for metadata and organization), and that location in the filesystem will stop being a useful organizational system.

I'll examine a few very common TB workflows, and show how they are handicapped by the lack of run metadata, and could be improved.

### The Workflow
I'm iterating on a model. Every time I tweak it, I re-run my model with new parameters (maybe once every minute or two with small models).

### Current Approaches
If I do this naïvely, then the output from my many "runs" will all wind up in a single directory (defined by re-creating the FileWriter with a fixed string); thus, the runs will be mangled together into one "run" that inappropriately concatenates the output from many TF runs. TensorBoard handles this badly, and will show "triangle charts" that finish, return to origin, and start again.

One less-naïve approach would be to write custom code for a FileWriter that auto-increments directory name, so I would wind up with "run1", "run2", "run3", etc. In practice, this would quickly lead to a very noisy TensorBoard in which the salient data (performance on my latest run) is obscured by the results from all my previous runs.

In practice, I simply delete all of my log data before each run, sidestepping the TB run model entirely. (In practice, this seems to occasionally break TB, so I need to restart it every few runs.)

### Better Approaches
Suppose we had the following three pieces of metadata for each run:
1. The time the run started
2. How successful the run was (i.e. a particular scalar summary is the key metric)
3. (ambitious) The code that generated the run

Then, we could have TensorBoard keep every run in the UI, but only show me my latest two runs, and my best run ever. Then I can glance at TB and immediately see how my new run is faring compared to my last approach, and my historical best. This would be particularly great on the scalar dashboard, since it would let me compare (for slow jobs) how good my run is at its latest step compared to my best ever run at the same step.

If we kept the code that generated the run, I could even do a code diff to remember what my approach was back when I had my most successful run. This would make TB into a much more effective "lab notebook" / "experiment journal". (On the same note, it would be awesome if I could add written annotations to each run.)

For this approach to work, it's assumed that every time the FileWriter/Run is initialized, it writes out a signature in the event log, so that we can distinguish between the last event of run `n-1` and the first event of run `n`.

### The Workflow
I have a model, and it's time to tune it via hyperparameter search. I define (say) 5 hps to vary, and 5 different values for each hp, resulting in 25 runs. I want to see how these different sets of HPs perform, with a particular emphasis on looking at the combinations that got the best performance.

### Current Approaches
I create a "hyperparameter string" for each run, e.g. "lr=1e-2,beta=0.5,n_fc=3,n_conv=4,act=relu". These strings need to fit in a small space in the UI, so interpretability is sacrificed. My only structured way to search across these approaches (e.g. to see how the class of models with learning rate 1e-2 and 4 convolutional layers performed) is to write regexes like "lr=1e-2.*n_conv=4". If my experiment generated a large number of runs, merely discovering which one performed best on my metrics may be a chore.

Here's an example (note: this TB has >100 runs, only a fraction shown)

![image](https://user-images.githubusercontent.com/1400023/30950015-cf6dcbe6-a3ce-11e7-8c10-d6943a4c9764.png)

### Better Approaches
We should add explicit support for hyperparameters with a flexible data model that includes:
* support for different kinds of hparams (binary, float val, integer val)
* provenance for hparams, since they may have come from intermediate tools/frameworks and be tricky to hunt down
* longform descriptions for hparams, so you can write down what they are

Then we can make a UI that has explicit support for this information, like:
* providing a [parallel coordinate plot](https://bl.ocks.org/jasondavies/1341281) that allows easy visualization of how different hparams affected performance
* surfacing the top-K runs according to a metric for explicit comparison
* UI affordance for filtering run selection by hyperparameters (perhaps drag on the parallel coordinates plot)

### The Workflow
Within each conceptual "run" as described above, we may record TF metrics on the training data, and then, at the end of the run, separately on a test set of data. The metric we care about may be performance on the test set rather than the train data.

### Current Apporaches
It's common to record separate runs for test and train by appending a "/train" or "/eval" suffix on the run name. Since TensorBoard is not aware that there is a relationship between "foo/train" and "foo/eval" but not between "foo/train" and "bar/eval", it can be very hard to filter them or compare them using the UI. Try to spot the correspondences between the train and eval-test versions of the same run in this example...

![image](https://user-images.githubusercontent.com/1400023/30950000-aaa862a8-a3ce-11e7-9248-b48d047056ef.png)

### Better Approaches
Since this is both very common and fundamental to machine learning, I think we should privilege it in the data model. Then, we can use the information to make more intelligent displays on the scalar dashboard, e.g. drawing all the eval losses and train losses on separate graphs, and using consistent color motifs for the train and eval portions of runs.

## Run organization
In addition to adding data that describes individual runs, I think we should add some higher level organization between runs. Especially since we expect to have organization-level TensorBoards that may serve dozens of users, each of whom have hundreds or thousands of runs!

A very simple system would be to have Projects as a top-level construct that defines boundaries of comparability. Two runs in different projects will never be relevant to compare, e.g. comparing an image model to a text model is meaningless. Within projects, experiments will contain tightly related runs, e.g. all the different runs for a given hyperparameter search.

However, it might be better to not make hard assumptions about the way that researchers will want to organize data, and instead come up with a simple abstraction they can apply to their purposes, e.g. a directory structure abstraction.

We should also add information like:
* owner
* group (esp. relevant for visibility restriction)

It might also be helpful to add "search terms" like "image", "mnist", "cifar", "gan", etc. This would assist people doing meta-analyses of research efforts, or looking to explore for successful approaches to related problems.

## Pitfalls
We need to make sure we can distinguish "started a new run" from "restarted a previous run from checkpoint" (we should be able to depend on the [SessionLog proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/event.proto#L64)).

## Field ideas
* Creation time
* Type (train/test/eval)
* Key metric id (/ performance of key metric?)
* Author
* Group
* ProjectId?
* Hyperparameters (likely use Tensor encoding for each HParam, to accomodate generic types)
* Longform description
* Annotations (generated via TB UI)
* Star (maybe useful so we can signal to TB to keep this run rather than garbage collecting - e.g. star just the best performing run in a hyperparameter search)
* UUID (useful for keeping in a large run DB)
* search keywords (should be modifiable from TB)

## Next Steps
Whoever takes this on (@wchargin?) should define a data model for runs that accomodates these use cases, and a corresponding API that lets users populate that data model, along with a sketch of how the new data will be used in the UI. Once we agree on the data model, come up with and execute a plan that gets these features into TensorBoard :) 

We don't need to add everything at once. An incremental approach would be fine.

cc @jart @chihuahua @nfelt

**Comment by chihuahua** [comment]
FYI, I am currently spec-ing out parallel coordinate plots with the internal team that does hyperparameter tuning, so I'll make sure to stay in the loop here.

**Comment by milani** [comment]
Since the last reference to this issue dates back to 6 months ago, I want to ask if this feature request is still relevant or the focus is shifted?

**Comment by decentralion** [comment]
@milani I was the big advocate for this, and I'm not working on TB anymore, so I think the focus has shifted. I'll close the issue
